[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99220

Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D45167585](https://our.internmc.facebook.com/intern/diff/D45167585)

jerryzh168 committed:
[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not

Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
> (2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

I dont follow this comment.

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -129,6 +130,31 @@
     "output_act_obs_or_fq_ctr": torch.ao.quantization.qconfig._default_quint8_placeholder_qconfig.activation
 }
 
+def _needs_obs_or_fq(prev_output_dtype, prev_output_is_dynamic, cur_target_dtype, cur_target_is_dynamic, reuse_input_obs_or_fq):
```
Will this account for something like
"
conv1
slice
view
conv2
"
Here the conv2's input and conv1's output can share observer such that conv1's output node will have quant/dequant op and conv2's input will also have quant/dequant op but their qparams are shared.

jerryzh168 replied:
yes I think so, it should be supported as long "slice" and "view" are added to the list in https://github.com/pytorch/pytorch/pull/99384

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -129,6 +130,31 @@
     "output_act_obs_or_fq_ctr": torch.ao.quantization.qconfig._default_quint8_placeholder_qconfig.activation
 }
 
+def _needs_obs_or_fq(prev_output_dtype, prev_output_is_dynamic, cur_target_dtype, cur_target_is_dynamic, reuse_input_obs_or_fq):
```
Can you just pass in the nodes?

jerryzh168 replied:
probably can't do this for output I think, since we didn't get `prev_output_dtype`, `prev_output_is_dynamic` from node.meta, it's hardcoded instead

kimishpatel replied:
I see

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -508,6 +534,40 @@ def _get_arg_target_dtype_as_output(
+    argument in quantized graph will match what is specified by the qconfig
+    """
+    assert isinstance(arg, Node)
+    # Custom module LSTM output is a tuple that we broke down into the internal nodes in order
+    # to insert DeQuantStubs (see `_insert_dequant_stubs_for_custom_module_lstm_output`).
+    # Since we modified the graph in this case, we must trace back from the args through
+    # the specific nodes we added in order to reach the original LSTM node. Otherwise, we would
+    # not be able to accurately detect whether this node is a consumer of custom module LSTM.
```
Why is this the case? And if this is because we inserted observers, then should be do node annotation so that you dont have to traverse the modified graph?

jerryzh168 replied:
you mean the LSTM related comment? this is for fx graph mode quantization

kimishpatel replied:
I think I understand now. The lstm part is unrelated to annotation and already has modified graph with dequant stub.

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -598,41 +658,53 @@ def _maybe_insert_input_observer_for_arg_or_kwarg(
+        act_post_process_ctr = target_dtype_info.get(obs_or_fq_ctr_key, _DEFAULT_FP32_OBS_OR_FQ_CTR)
 
         arg_as_output_target_dtype = _get_arg_target_dtype_as_output(arg, named_modules)
+        arg_as_output_target_is_dynamic = _get_arg_target_is_dynamic_as_output(arg, named_modules)
```
So real special case here is lstm where we need to trace back through dequant stub to find the node? Is this to preserve existing code for fx quant? and will this be relevant for new workflow?

jerryzh168 replied:
yeah not relevant for new flow, that code path is not used

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -805,20 +877,43 @@ def _maybe_insert_output_observer_for_node(
+    # uncomment after we support reuse_input_obs_or_fq properly by having separate
+    # implemntations for this key instead of reusing the input_output_share_observers
+    # code
+    # reuse_input_obs_or_fq = node.meta["target_dtype_info"].get("reuse_input_obs_or_fq", False)
```
but what is the difference between observer sharing and `reuse_input_obs_or_fq`

jerryzh168 replied:
they might be the same, I can follow up to see if I can merge them

kimishpatel commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -805,20 +877,43 @@ def _maybe_insert_output_observer_for_node(
+    # because the prev_output is the output of an fp32 op, althought technically
+    # we should get the dtype of the output from node.meta["val"] in the future
+    # if we deprecate fx graph mode quantization
+    needs_obs_or_fq = _needs_obs_or_fq(torch.float, False, target_dtype, target_is_dynamic, reuse_input_obs_or_fq)
```
THis is just saying that output observer is always needed unless target_dtype == float?

jerryzh168 replied:
I think so, we use the same logic for inserting input observer here

kimishpatel reviewed (changes requested):
Left some comments, but I am wondering if we can avoid these complexity by only tagging output of the nodes like we discussed yesterday. Let me share a PR I have that just tags `output_act_obs_or_fq_ctr` for linear nodes and see if that makes sense and if there are corner cases I missed. With that approach, I think it would make more sense for us to focus on how to remove redundant observers or fq nodes.

jerryzh168 commented:
> > (2) output for conv node
> > the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
> > is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
> > current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
> > after we decide to deprecate fx graph mode quantization
> 
> the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]
> 
> I dont follow this comment.

this is talking about the rule for output of a node, we will use the same rule in `_needs_obs_or_fq` to decide whether to insert an observer or not and the `prev_output_dtype`, `prev_output_is_dynamic` are hardcoded to torch.float, False right now.

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -129,6 +130,31 @@
+    if cur_target_dtype in _OBS_DTYPE_LIST:
+        return prev_output_dtype in _OBS_DTYPE_LIST + [torch.float] and cur_target_dtype != prev_output_dtype
+
+    # lots of error checking are skipped here for now
```
mostly for user specifies inconsistent things, see the table in design doc for detail

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -129,6 +130,31 @@
+        return False
+    # non dynamic quantization
+    if cur_target_dtype in _OBS_DTYPE_LIST:
+        return prev_output_dtype in _OBS_DTYPE_LIST + [torch.float] and cur_target_dtype != prev_output_dtype
```
we don't insert extra observers if we go from torch.quint8 -> torch.float
also we don't insert observers if prev_output_dtype is None, int, float, bool etc.
some of these can potentially be simplified I think, but this PR more of a cleanup of the current logic to make it clearer and also enable the "Not Specified" behavior

kimishpatel replied:
> we don't insert extra observers if we go from torch.quint8 -> torch.float

This is just an artifact of how we are doing this today. If we have no default confict with dtype = float, we should even worry about this, right? 

Can we file a task for this to be tracked?

kimishpatel replied:
> insert observer both when we go from torch.quint8 -> torch.float and torch.float -> torch.quint8

This makes sense.

I was just highlighting that we `prev_output_dtype ! = cur...` and stuff is there due to how we are annotating target_dtype_ stuff with input, weight, output, bias etc. I suspect if we did only output based stuff (while resolving the issue of how to insert different observers for different "uses").

jerryzh168 replied:
sounds good, discussed offline that we would like to refactor the annotation API to annotate "use" instead of nodes, will do it in a separate PR

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -598,41 +658,53 @@ def _maybe_insert_input_observer_for_arg_or_kwarg(
         # for nodes that doesn't have `reuse_input_obs_or_fq` configured,
         # we'll default to False, this makes configuring this field optional for users
-        reuse_input_obs_or_fq = node.meta["target_dtype_info"].get("reuse_input_obs_or_fq", False)
+        reuse_input_obs_or_fq = target_dtype_info.get("reuse_input_obs_or_fq", False)
```
yeah they might be the same I think, but need to verify

kimishpatel replied:
Lets file a task for this as well.

jerryzh168 replied:
done: https://github.com/pytorch/pytorch/issues/99636

kimishpatel replied:
I have also created internal tasks. Lets link this issue with that.

jerryzh168 commented:
> Left some comments, but I am wondering if we can avoid these complexity by only tagging output of the nodes like we discussed yesterday. Let me share a PR I have that just tags `output_act_obs_or_fq_ctr` for linear nodes and see if that makes sense and if there are corner cases I missed. With that approach, I think it would make more sense for us to focus on how to remove redundant observers or fq nodes.

I feel it might make more sense to get this working with the current code base first, since what we discussed yesterday require major refactorings and making sure bc for existing models etc. I think that is a discussion that can happen later. For now I feel you can do:
```
match = get_match(....)
weight_node = match.weight_node
transpose_node = weight_node.users[0]
transpose_node.meta["target_dtype_info"]["input_act_obs_or_fq"] = weight_obs_ctr

...
```

and in the future if we decide to remove 'input_act_obs_or_fq' after evaluating the benefits of cost of the change, it would just be changing from the above to:
```
match = get_match(....)
weight_node = match.weight_node
weight_node.meta["target_dtype_info"]["output_act_obs_or_fq"]
```

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
> match = get_match(....)
> weight_node = match.weight_node
> weight_node.meta["target_dtype_info"]["output_act_obs_or_fq"] = weight_obs_ctr

This is how I have done in #99580 and it seems to be working although I handled tranpose node differently. But not doing anything about it but tagging as annotated.

kimishpatel commented:
Lets follow up on #99580 and figure out that this PR and the other one can play nice with each other. then we can land this.

If we do land this, I would like us to file a couple of follow up tasks that were mentioned in the comments.

jerryzh168 commented:
> > match = get_match(....)
> > weight_node = match.weight_node
> > weight_node.meta["target_dtype_info"]["output_act_obs_or_fq"] = weight_obs_ctr
> 
> This is how I have done in #99580 and it seems to be working although I handled tranpose node differently. But not doing anything about it but tagging as annotated.

Yes, the current code works for the use case already, although with the caveat that you'll need to duplicate all nodes that's used multiple times, or add identity node for them to actually make it correct, this the annotation in the PR does not work if a node is used by multiple nodes right now

kimishpatel commented:
> you'll need to duplicate all nodes that's used multiple times, or add identity node and also add a dedup observers pass

What would be the follow up from this?

jerryzh168 commented:
> > you'll need to duplicate all nodes that's used multiple times, or add identity node and also add a dedup observers pass
> 
> What would be the follow up from this?

I think we should lay out the benefits and cost to decide which one we want to use, would also be great to be backed by concrete examples. so would be nice if we onboarded few models with the pattern matcher so we have some concrete annotations (the assumption of one user might be true for most of the models I guess)

kimishpatel commented:
> maybe except for residual connection for resnet

THis is a good one to track but I dont know if we are actually quantizing the two usages differnetly anyway

jerryzh168 commented:
> > 
> 
> This is independent of pattern matching vs module partition, right? Just want to make sure we are talking about the same thing.

it's related, I think the main issue you are talking about is, since we are using pattern matching now, we want to annotate the input/output nodes only, right?

jerryzh168 commented:
> > maybe except for residual connection for resnet
> 
> THis is a good one to track but I dont know if we are actually quantizing the two usages differnetly anyway

yeah no matter how we quantize them it would be better to actually try out this and see what are the problems that need to be solved, right now I don't feel it is better than the existing annotation API

kimishpatel commented:
> > > 
> > 
> > 
> > This is independent of pattern matching vs module partition, right? Just want to make sure we are talking about the same thing.
> 
> it's related, I think the main issue you are talking about is, since we are using pattern matching now, we want to annotate the input/output nodes only, right?

I was referring to using only output node annotation instead (input, output, weight, bias) etc.

jerryzh168 commented:
> 
> I was referring to using only output node annotation instead (input, output, weight, bias) etc.

right, I understand, I think we can try out this on some models ops and see what is needed to make it work e2e before deciding, things that should be needed are duplicate nodes before observation and dedup observers after quantization (the logic of dedup observers, is doing this based on dtype enough? etc.) need to be resolved, and we can discuss the cost and benefits afterwards

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e][refactor] Cleanup the logic for deciding whether to insert observer/fq or not"


Summary:
Previously we have two places we need to decide whether to insert observer or fake quantizer or not:
(1) input arguments of a node (2) output of a node, and right now we have separate code to do this
in this PR, the logic is unified in `_needs_obs_or_fq` helper function that takes the target_dtype and is_dynamic from previous output
and target_dtype and is_dynamic for the current Tensor we are looking at

let's use an example for conv node:
```
conv = convolution(input, weight, bias, ...)
```

let's say we have `input_node` object for argument `input`, and `conv_node` for `conv` node in the graph

(1) input arguments, e.g. `input`
the target_dtype/is_dynamic from previous output is the node that produces `input`, we get this from
input_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

the taregt_dtype/is_dynamic for the current argument `input`, comes from conv_node.meta["target_dtype_info"]["input_act_obs_or_fq"]
similarly for weight it comes from conv_node.meta["target"]["weightobs_or_fq"] etc.

(2) output for conv node
the target_dtype/is_dynamic from previous output will be the floating point output from the fp32 convolution operator, so it
is hardcoded to be (torch.float, False), however, technically we should get this from node.meta["val"], but since the
current code base is shared by fx graph mode quantization and pytorch 2.0 export quantization, we cannot do that, we can revisit
after we decide to deprecate fx graph mode quantization

the target_dtype/is_dynamic for the current output comes from conv_node.meta["target_dtype_info"]["output_act_obs_or_fq"]

there is one caveat here about dynamic quantization, that is explained in the comment, so I won't repeat here

Note: also fixed some places in `_get_arg_target_dtype_as_input_to_node` and `_get_arg_target_is_dynamic_as_input_to_node` to make sure "not specified" == specifying a fp32 placeholder observer as well

Next: we can merge the two get target dtype and get is_dynamic function to reduce code duplication

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
python test/test_quantization.py TestQuantizePT2E
python test/test_quantization.py TestQuantizePT2EModels

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D45167585).

kit1980 commented:
Several lint failures, need to replace the calls the removed function.

 Error (MYPY) [name-defined]
    Name "_get_default_obs_or_fq_ctr" is not defined
    
https://github.com/pytorch/pytorch/actions/runs/4767109379/jobs/8474994139

kit1980 commented:
@pytorchbot revert -m "Lint caught several issues" -c nosignal
